### ISSUE TITLE
PositN(pi)  MethodError: no method matching Posit32(::Irrational{:π})

### ISSUE THREAD

**Issue by tkgunaratne** [takum-arithmetic/Posits.jl#3]
Hi @hunhold,

Thanks for this Package!! I just wanted to pointed out PositN(pi) for N=8, 16+, 32, 64 results a method error. However, PositN(1*pi) works! 

I'm pleasantly surprised that with your implementation of Posits, I managed to evaluate FFT/IFFT for non power of 2 points. This is a feature I have been looking for a while for my DSP applications.

Regards,

tkg.

**Comment by hunhold** [comment]
Dear @tkgunaratne,

this is a really good find, thank you for reporting this! I am just a bit puzzled on how to best approach this. Do you think it's sufficient to define e.g.

`Posit8(i::Irrational) = Posit8(Float64(i))`

or does it require more work?

I am glad to hear that you find this library useful. In which precision did you evaluate posits regarding FFT/iFFT?

**Comment by tkgunaratne** [comment]
Hi @hunhold,

Thanks for your quick response. I'm afraid that I'm not a Julia expert to make suggestions on how to address this. 

" In which precision did you evaluate posits regarding FFT/iFFT?" 
For my applications, I need SNR better than 60 dB. My initial tendency to use is Posit16 because it would be closer to the data widths that I'm currently using. However for FFTs of order 2^14, the resulting SNR is not enough so, I might have to use Posit32 for the implementation of FFTs and then convert back to Posit16.

tkg.

My code for FFT/IFFT comparison with Posits and Floats:
```
#This Julia script test the the implementation of the fft/ifft with posits
using Posits;
using FastTransforms;
using Statistics;

P0 = Posit16;
P1 = Posit32;

# P0 = Float32;
# P1 = Float32;

# P0 = Float32;
# P1 = Float32;

#Number of fft/ifft points
#n = 2^14;
n = 20;

a_f = 1*(randn(n) + im*randn(n))/sqrt(2);
a_p = P0.(real(a_f)) + im*P0.(imag(a_f));

#Initial Conversion Error
a_r = Float64.(real(a_p)) + im*Float64.(imag(a_p));
Error_f = (a_f - a_r);
Max_E_f = maximum(abs.(Error_f));
SNR_E_f = 10*log10(var(a_r)/var(Error_f))
println("Maximum Error in Data $Max_E_f")
println("The SNR of the Quantized Data $SNR_E_f dB\n") 

#Converting the Data to evalaute the fft
a_p1 = P1.(real(a_p)) + im*P1.(imag(a_p));
#Forward fft
@time A_F = fft(a_f);
@time A_P = fft(a_p1);
A_R = Float64.(real(A_P)) + im*Float64.(imag(A_P));

#Post FFT Error
Error_F = (A_F - A_R);
Max_E_F = maximum(abs.(Error_F));
SNR_E_F = 10*log10(var(A_R)/var(Error_F))
println("\nMaximum Error in FFT $Max_E_F")
println("The SNR of the Quantized FFTs $SNR_E_F dB\n")

**Comment by hunhold** [comment]
Dear @tkgunaratne,

I will put the pi-issue on the TODO; it should be easy to fix.

Regarding FFT/iFFT: Have you tried if takums work better for you? There is a package called "Takums", which I also develop. Testing them is as simple as "using Takums" and replacing PositN with LinearTakumN. I am looking forward to hear back from you!

**Comment by tkgunaratne** [comment]
I had previously glanced through your article "[Beating Posits at Their Own Game: Takum Arithmetic](https://arxiv.org/abs/2404.18603)" and was glad to be able to compare it with other formats. Here are my initial results for evaluating  16,384-point FFT with Float64 and 16 and 32-bit LinearTakums, Poisits and Floats. I tried the FFT with 16-bit equivalents but the resulting SNR is well below 60 dB.
```
Quantized Data with LinearTakum16:
Maximum Error in Data 0.0009788458294267212
The SNR of the Quantized Data 78.13691880736906 dB

 FFT Float64 - 0.005225 seconds (447 allocations: 286.344 KiB, 82.97% compilation time)
 FFT  LinearTakum32 1.698514 seconds (1.96 M allocations: 134.280 MiB, 2.26% gc time, 77.22% compilation time)

Maximum Error in 16384-point FFT 0.04811441257195696
With LinearTakum32 Arithmetics, The SNR of the Quantized FFTs 78.13663072987298 dB

Quantized Data with Posit16:
Maximum Error in Data 0.0005270731115253108
The SNR of the Quantized Data 79.70284521110385 dB

  FFT Float64 0.000685 seconds (6 allocations: 256.344 KiB)
  FFT Posit32 0.988081 seconds (664.04 k allocations: 44.960 MiB, 13.09% gc time, 61.07% compilation time)

Maximum Error in 16384-point FFT 0.04071731043780854
With Posit32 representation, the SNR of the Quantized FFTs 79.70205362075734 dB

Quantized Data with Float16:
Maximum Error in Data 0.001019620424992161
The SNR of the Quantized Data 73.68785662111732 dB

  FFT Float64 0.000487 seconds (6 allocations: 256.344 KiB)
  FFT Float32 0.004187 seconds (442 allocations: 157.953 KiB, 88.40% compilation time)

Maximum Error in 16384-point FFT 0.0839663063677031
With Float32 representation, the SNR of the Quantized FFTs 73.688134599693 dB

**Comment by hunhold** [comment]
Dear @tkgunaratne, thanks again for mentioning this issue. I added the irrational conversions (#e2a27b4) and they will be part of the next release.

Regarding your code example: During my own tests, I noticed that there is a high variance in the test results. I recommend running at least 5 (better much more until the data is good) trials and determining the mean value and standard deviation. Feel free to send me an e-mail for further discussion.
